### ISSUE TITLE
Move to ghc-lib-parser phase2

### ISSUE THREAD

**Issue by ndmitchell** [ndmitchell/hlint#892]
Ultimate goal is to drop the HSE dependency. Following on from #645. Tasks left:

- [x] Delete and deprecate Config.Haskell files. They may well not work anymore anyway and Yaml has been recommended for a while.
- [x] Port over Config.Compute - I'd be sad to see this disappear.
- [x] A few bits and pieces in Apply use hseModule for order and classification
- [x] Decide what to do about the Grep module. Does anyone use it? Should we just kill it?
- [x] Things like Test.Translate haven't worked in a while are based on the old code. Comment out and add an error saying if you want it, reenable it?
- [x] Test.Proof needs dealing with. Same approach as Test.Translate?
- [x] Idea needs to not use HSE SrcSpans
- [x] Get rid of ParseFlags
- [x] Convert data/Test.hs to test.yaml (moved to #932)

CC @googleson78 and @shayne-fletcher-da

**Comment by shayne-fletcher** [comment]
Looking forward to continuing the work. Is a replacement for HSE's `parseFlags` mechanism covered in the above? I think getting that in will help drive us along some.

**Comment by ndmitchell** [comment]
Wasn't aware parseFlags was required. I can't find it in the code base now.

**Comment by shayne-fletcher** [comment]
See `HSE/All.hs`. Everything is driven off `ParseFlags/ParseMode`.

**Comment by shayne-fletcher** [comment]
> Wasn't aware parseFlags was required. I can't find it in the code base now.

Sorry, spelling mistake that I'm sure threw you off.

**Comment by shayne-fletcher** [comment]
Ok, quick look at Hoogle suggests that it might not prove too onerous.

**Comment by ndmitchell** [comment]
Doh, I had my grep in case-sensitive mode. Agreed, that needs ripping out - I've added it to the list. Unlike phase1, I think these are all "can be done in an afternoon" tasks, so hopefully a lot less onerous.

**Comment by shayne-fletcher** [comment]
Consider me assigned to that one next when it's in the list if it suits you.

**Comment by googleson78** [comment]
I did.. some work(?) on `Idea needs to not use HSE SrcSpans` - #894

**Comment by ndmitchell** [comment]
Thanks @googleson78 - that's now landed. As a consequence, the HLint4 API has gone from exposing HSE.SrcSpan to GHC.SrcSpan. That's a breaking change, and so as to ensure we only have one breaking API change, it's better to switch _all_ HSE types to GHC types in the next release so that we can then eliminate HSE without a further API break.

I've just pushed most of the little things. The ones left are Fixity and ParseMode. I think we should lift most of the fields we care about in ParseMode directly into ParseFlags, e.g. have parseFixities :: [Fixity], parseExtensions :: [Extension] and whatever else we think is useful. Once that's all done, we're in a position to release once more, with a corresponding 0.1 API breaking bump.

**Comment by shayne-fletcher** [comment]
Earliest I can get to resuming is the weekend (swamped and got a bit of burn-out going on). If that's a problem and it needs to be pushed on by someone else, I won't be offended.

**Comment by ndmitchell** [comment]
If you've got a bit of burn-out going on, can I suggest you _don't_ get to this at the weekend, but take the weekend to take the dog for a walk? There's no rush on this.

**Comment by shayne-fletcher** [comment]
You guys are the best ❤️

**Comment by ndmitchell** [comment]
OK, we're down to ParseFlags as the last outstanding thing that needs solving before a release. No rush on it though.

**Comment by shayne-fletcher** [comment]
PR https://github.com/ndmitchell/hlint/pull/900 makes a start on parse flags and fixities.

**Comment by shayne-fletcher** [comment]
PR https://github.com/ndmitchell/hlint/pull/924 removes `haskell-src-exts` from `hlint.cabal`.

**Comment by ndmitchell** [comment]
I deleted the old API's at HLint and HLint3, and moved over HLint4 to be HLint, so now we have the nice API we've always wanted in terms of GHC constructs. I went through the API, improved it in a few little places (merging ParseMode and ParseFlags), and I'm now happy with it. The only thing still on the list was converting a test, so I've spun that out as #932 and we can close this issue.

Since we are now tied to GHC, each GHC release has a strong likelihood of breaking the API for everyone. Therefore, I'm going to hold off on a release until we either upgrade to 8.10 as per #927, or declare that such an upgrade will take a week or so, and then release immediately.
